controllers/medusa/medusatask_controller.go

```diff
@@ -0,0 +1,470 @@
+	err := r.Get(ctx, req.NamespacedName, instance)
+	if err != nil {
+		logger.Error(err, "Failed to get MedusaTask")
+		if errors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{RequeueAfter: r.DefaultDelay}, err
+	}
+
+	task := instance.DeepCopy()
+
+	// First check to see if the task is already in progress
+	if !task.Status.StartTime.IsZero() {
+		// If there is anything in progress, simply requeue the request
+		if len(task.Status.InProgress) > 0 {
+			logger.Info("Tasks already in progress")
+			return ctrl.Result{RequeueAfter: r.DefaultDelay}, nil
+		}
+
+		logger.Info("task complete")
+
+		// Set the finish time
+		// Note that the time here is not accurate, but that is ok. For now we are just
+		// using it as a completion marker.
+		patch := client.MergeFrom(task.DeepCopy())
+		task.Status.FinishTime = metav1.Now()
+		if err := r.Status().Patch(ctx, task, patch); err != nil {
+			logger.Error(err, "failed to patch status with finish time")
+			return ctrl.Result{RequeueAfter: r.DefaultDelay}, err
+		}
+
+		// Schedule a sync if the task is a purge
+		if task.Spec.Operation == medusav1alpha1.OperationTypePurge {
+			r.scheduleSyncForPurge(task)
+		}
+
+		return ctrl.Result{Requeue: false}, nil
+	}
+
+	// If the task is already finished, there is nothing to do.
+	if taskFinished(task) {
+		logger.Info("Task is already finished", "MedusaTask", req.NamespacedName, "Operation", task.Spec.Operation)
+		return ctrl.Result{Requeue: false}, nil
+	}
+
+	// Task hasn't started yet
+	cassdcKey := types.NamespacedName{Namespace: task.Namespace, Name: task.Spec.CassandraDatacenter}
+	cassdc := &cassdcapi.CassandraDatacenter{}
+	err = r.Get(ctx, cassdcKey, cassdc)
+	if err != nil {
+		logger.Error(err, "failed to get cassandradatacenter", "CassandraDatacenter", cassdcKey)
+		return ctrl.Result{RequeueAfter: r.DefaultDelay}, err
+	}
+
+	pods, err := r.getCassandraDatacenterPods(ctx, cassdc, logger)
+	if err != nil {
+		logger.Error(err, "Failed to get datacenter pods")
+		return ctrl.Result{RequeueAfter: r.DefaultDelay}, err
+	}
+
+	if task.Spec.Operation == medusav1alpha1.OperationTypePurge {
+		return r.purgeOperation(ctx, task, pods, logger)
+	} else if task.Spec.Operation == medusav1alpha1.OperationTypeSync {
+		return r.syncOperation(ctx, task, pods, logger)
+	} else if task.Spec.Operation == medusav1alpha1.OperationTypePrepareRestore {
+		return r.prepareRestoreOperation(ctx, task, pods, logger)
+	} else {
+		return ctrl.Result{}, fmt.Errorf("unsupported operation %s", task.Spec.Operation)
+	}
+}
+
+func (r *MedusaTaskReconciler) getCassandraDatacenterPods(ctx context.Context, cassdc *cassdcapi.CassandraDatacenter, logger logr.Logger) ([]corev1.Pod, error) {
+	podList := &corev1.PodList{}
+	labels := client.MatchingLabels{cassdcapi.DatacenterLabel: cassdc.Name}
+	if err := r.List(ctx, podList, labels); err != nil {
+		logger.Error(err, "failed to get pods for cassandradatacenter", "CassandraDatacenter", cassdc.Name)
```

Comment:
In general I don't think we should log an ERROR statement when returning an error. It can create noise in the logs. See #281 for more info. If we need to log something it is probably for debugging purposes. See #282 for details.

Reply:
I saw you wanted us to remove these loggings. The problem is that the error printed out if we don't put our own logging is usually more obscure and doesn't give details allowing to know precisely what failed. Technically we know what failed, but it's harder to know what we were trying to do when it failed.
I'm ok with some repetition in the logs if that allows more clarity during troubleshooting.

Reply:
The error printed out should include your error message. What isn't included though is a stack trace which can be problematic when the same error message is used in multiple places. 

Whatever we do or change (if anything) for logging is a separate concern from this PR, so I think we're good to go.